keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,348 @@
+
+- Logs for deployed pods related to the KUDO Instance
+- YAML for created resources, including both spec and status, related
+  to the KUDO instance
+- Output of `kubectl describe` for all deployed resources related to the KUDO
+  Instance
+- Current plan status, if one exists, or the KUDO Instance
+- Information about the KUDO instance's Operator and OperatorVersion
+- Logs for the KUDO controller manager
+- Describe for the KUDO controller manager resources
+- RBAC resources that are applicable to the KUDO controller manager
+- Current settings and version information for KUDO
+
+By default, KUDO scrubs all resources (and custom resources) of values that are
+contained within the KUDO Instance's secrets. This is configurable with the
+`diagnostics.scrubSecrets` key.
+
+Operator developer experience, then, focuses on customizing diagnostics
+information to gather information about the running application. The following
+forms are available, subject to change over time:
+
+- **File**: Copy a file out of a running pod. This is useful for non-stdout
+  logs, configuration files, and other artifacts generated by an application.
+  Higher level resources can also be used, which will copy the file on all pods
+  selected by that resource.
+- **Command**: Run a command on a running pod and copy the stdout. Higher level
+  resources can also be used, which will run the command on all pods selected by
+  that resource.
+- **Job**: Run a job and copy the stdout and other arbitrary files.
+- **HTTP**: Make an HTTP request from the KUDO controller manager to a named
+  service and port and copy the result of the request.
+- **Scrub**: Scrubs all collected resources, including default resources, of a
+  particular value. This can be a string, regular expression, or
+  JSONPath/objectRef combination from the KUDO instance's resources. Generally,
+  these should be the last steps if used.
+
+While some of these are redundant (HTTP can be a command or job), the intent
+is to provide a high level experience where possible so that operator developers
+don't necessarily need to maintain a `curl` container as part of their
+application stack.
+
+Operator-defined diagnostics collection is defined in a new `diagnostics.bundle`
+key in `operator.yaml`:
+
+```yaml
+diagnostics:
+  bundle:
+    - name: Zookeeper Configuration File
+      key: "zookeeper-configuration"
+      kind: File
+      spec:
+        path: /opt/zookeeper/server.properties
+        objectRef:
+          kind: StatefulSet # Runs on ALL pods in the statefulset
+          name: "{{ .InstanceName }}-zookeeper"
+    - name: DNS information for running pod
+      kind: Command
+      spec:
+        command: # Can be string or array
+          - nslookup
+          - google.com
+        objectRef:
+          kind: Pod
+          name: "{{ .InstanceName }}-zookeeper-0"
+    - name: Scrub authentication information
+      kind: Scrub
+      spec:
+        regex: "/^host: %w+$/"
+```
```

Comment:
Do we need an `output` attribute? I.e. how to name the file where the output is written, or can we derive that from the rest of the attributes?

Reply:
Yes, this would not be correct since it doesn't have `key` (described below).

Reply:
Oh, no, not for scrub. Scrub is scrubbing existing files.

Reply:
Maybe scrub needs to be moved to it's own key then!

Reply:
updated to that, I think it ends up semantically much better.